cdap-master/src/main/java/co/cask/cdap/data/tools/ReplicationStatus.java

```diff
@@ -0,0 +1,290 @@
+import co.cask.cdap.data2.dataset2.lib.table.hbase.HBaseTable;
+import co.cask.cdap.data2.util.TableId;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtilFactory;
+import co.cask.cdap.data2.util.hbase.ScanBuilder;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellUtil;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.client.HTable;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The Replication Status tool is used to track the status of replication from Master to a Slave Cluster.
+ * The tool reads writeTime and replicateTime for all tables and uses it to determine if all CDAP tables have been
+ * completely replicated to the Slave cluster.
+ * */
+public class ReplicationStatus {
+
+  private static int writeTime = 0;
+  private static int replicateTime = 1;
+
+  private static CConfiguration cConf = CConfiguration.create();
+  private static Configuration hConf = HBaseConfiguration.create();
+
+  protected static final Logger LOG = LoggerFactory.getLogger(ReplicationStatus.class);
+
+  private static String masterCluster = null;
+  public static void main (String args[]) throws Exception {
+
+    if (args.length < 1 || args[0].equals("help")) {
+      System.out.printf("Tool to check replication status to Slave Cluster.\n\n");
+      System.out.printf("Run this tool on the Slave cluster as:\n\n");
+      System.out.println("#ReplicationStatus <master cluster id:>");
+      return;
+    } else if (args[0].equals("debugdump")) {
+      dumpReplicationStateTable();
+    } else {
+      masterCluster = args[0];
+      checkHBaseReplicationComplete();
+    }
+  }
+
+  private static void checkHBaseReplicationComplete() throws Exception {
+
+    HBaseTableUtil tableUtil = new HBaseTableUtilFactory(cConf).get();
+    HTable hTable = tableUtil.createHTable(hConf, getReplicationStateTableId(tableUtil));
+    long shutdownTime = 0L;
+
+    // Scan the table to scan for all regions.
+    ScanBuilder scan = getScan(tableUtil);
+    Result result;
+
+    HashMap<String, ArrayList<Long>> regionMap = new HashMap<>();
+    try (ResultScanner resultScanner = hTable.getScanner(scan.build())) {
+      while ((result = resultScanner.next()) != null) {
+        ByteBuffer row = ByteBuffer.wrap(result.getRow());
+        String clusterID = new String(getNextPart(row), "UTF-8");
+        String nextPart = new String(getNextPart(row), "UTF-8");
```

Comment:
It would be cleaner to have a ReplicationStatusKey object that has clusterID, region, and whatever other fields the key is built from. That class should have a method to encode it into a byte[], and a method to decode it from a byte[]. That way encoding, decoding logic is in one class instead of spread out across the coprocessor and this tool